torch/onnx/_internal/_beartype.py

```diff
@@ -0,0 +1,94 @@
+"""An internal wrapper for the beartype library.
+
+The module returns a no-op decorator when the beartype library is not installed.
+"""
+import functools
+import traceback
+import warnings
+
+from torch.onnx import errors, _exporter_states
+from torch.onnx._globals import GLOBALS
+
+
+def _no_op_decorator(func):
+    return func
+
+
+if GLOBALS.runtime_type_check == _exporter_states.RuntimeTypeCheckState.DISABLED:
+    # Return a simple no-op decorator when TYPE_CHECKING to make mypy happy
+    beartype = _no_op_decorator
+elif GLOBALS.runtime_type_check == _exporter_states.RuntimeTypeCheckState.ENABLED:
+    # Enable runtime type checking which errors on any type hint violation.
+    try:
+        from beartype import roar as _roar
+        import beartype as _beartype_lib
+
+        # Beartype warns when we import from typing because the types are deprecated
+        # in Python 3.9. But there will be a long time until we can move to using
+        # the native container types for type annotations (when 3.9 is the lowest
+        # supported version). So we silence the warning.
+        warnings.filterwarnings(
+            "ignore",
+            category=_roar.BeartypeDecorHintPep585DeprecationWarning,
+        )
+        beartype = _beartype_lib.beartype
+    except ImportError:
+        # If the beartype library is not installed, return a no-op decorator
+        warnings.warn(
+            "TORCH_ONNX_EXPERIMENTAL_RUNTIME_TYPE_CHECK is set to '1', "
+            "but the beartype library is not installed."
+            "Install beartype with `pip install beartype` to enable runtime type checking."
+        )
+        beartype = _no_op_decorator
+else:
+    # Warning only
+    try:
+        from beartype import roar as _roar
+        import beartype as _beartype_lib
+
+        # Beartype warns when we import from typing because the types are deprecated
+        # in Python 3.9. But there will be a long time until we can move to using
+        # the native container types for type annotations (when 3.9 is the lowest
+        # supported version). So we silence the warning.
+        warnings.filterwarnings(
+            "ignore",
+            category=_roar.BeartypeDecorHintPep585DeprecationWarning,
+        )
+
+        def beartype(func):
+            """Warn on type hint violation."""
+
+            if "return" in func.__annotations__:
+                # Remove the return type from the func function's
+                # annotations so that the beartype decorator does not complain
+                # about the return type.
+                return_type = func.__annotations__["return"]
+                del func.__annotations__["return"]
+                beartyped = _beartype_lib.beartype(func)
+                # Restore the return type to the func function's annotations
+                func.__annotations__["return"] = return_type
```

Comment:
Is it ok for this to be here instead of the `finally:` block?

Reply:
I will need to verify

Reply:
Added a test